### ISSUE TITLE
[CI] AsyncSearchActionIT.testUpdateStoreKeepAlive fails

### ISSUE THREAD

**Issue by tlrx** [elastic/elasticsearch#67092]
The test `AsyncSearchActionIT.testUpdateStoreKeepAlive` failed today on 7.x.

**Build scan**:
https://gradle-enterprise.elastic.co/s/ee56ukpaummx4

**Repro line**:
```
./gradlew ':x-pack:plugin:async-search:internalClusterTest' --tests "org.elasticsearch.xpack.search.AsyncSearchActionIT.testUpdateStoreKeepAlive" -Dtests.seed=44FC88CE59837C9C -Dtests.security.manager=true -Dtests.locale=is -Dtests.timezone=Asia/Tehran -Druntime.java=8
```

**Reproduces locally?**:
No

**Applicable branches**:
7.x

**Failure history**:
Failed once today.

**Failure excerpt**:
```
java.util.concurrent.ExecutionException: RemoteTransportException[[node_s0][127.0.0.1:38963][indices:data/read/async_search/get]]; nested: RemoteTransportException[[node_s1][127.0.0.1:43749][indices:data/read/async_search/get]]; nested: ResourceNotFoundException[FmlmQVI5RUlmVFZlSEhqR2tRbFpyaGcaMTNUM21sY2lTeEtYZGNDVzZUN1IxUToyMjI=];
        at __randomizedtesting.SeedInfo.seed([44FC88CE59837C9C:7D414CB2B4D0C3F7]:0)
        at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:273)
        at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:260)
        at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:87)
        at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase.getAsyncSearch(AsyncSearchIntegTestCase.java:170)
        at org.elasticsearch.xpack.search.AsyncSearchActionIT.testUpdateStoreKeepAlive(AsyncSearchActionIT.java:400)

        Caused by:
        RemoteTransportException[[node_s0][127.0.0.1:38963][indices:data/read/async_search/get]]; nested: RemoteTransportException[[node_s1][127.0.0.1:43749][indices:data/read/async_search/get]]; nested: ResourceNotFoundException[FmlmQVI5RUlmVFZlSEhqR2tRbFpyaGcaMTNUM21sY2lTeEtYZGNDVzZUN1IxUToyMjI=];

            Caused by:
            RemoteTransportException[[node_s1][127.0.0.1:43749][indices:data/read/async_search/get]]; nested: ResourceNotFoundException[FmlmQVI5RUlmVFZlSEhqR2tRbFpyaGcaMTNUM21sY2lTeEtYZGNDVzZUN1IxUToyMjI=];

                Caused by:
                ResourceNotFoundException[FmlmQVI5RUlmVFZlSEhqR2tRbFpyaGcaMTNUM21sY2lTeEtYZGNDVzZUN1IxUToyMjI=]
                    at org.elasticsearch.xpack.core.async.AsyncTaskIndexService.lambda$getEncodedResponse$6(AsyncTaskIndexService.java:277)
                    at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:128)
                    at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43)
                    at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:94)
                    at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:88)
                    at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:251)
                    at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:242)
                    at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1289)
                    at org.elasticsearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:253)
                    at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:245)
                    at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:133)
                    at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:89)
                    at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:700)
                    at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142)
                    at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117)
                    at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82)
                    at org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:296)
                    at org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:228)
                    at org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:40)
                    at org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:139)
                    at org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:151)
                    at org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:420)
                    at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:246)
                    at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:174)
                    at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:131)
                    at java.lang.Thread.run(Thread.java:748)
```

**Comment by valeriy42** [comment]
Another failures:
* https://gradle-enterprise.elastic.co/s/4i6g2p5ldnwmk
* https://gradle-enterprise.elastic.co/s/ie3tcmle7kg6c

**Comment by pugnascotia** [comment]
Another failure but a different message - https://gradle-enterprise.elastic.co/s/hlbba7jpfiwm6

```
Expected: a value greater than <1638199886894L>
    but: <1638199886894L> was equal to <1638199886894L>
```

**Comment by dimitris-athanasiou** [comment]
And another failure: https://gradle-enterprise.elastic.co/s/6odmjpmerlk7a

**Comment by andreidan** [comment]
Another failure: https://gradle-enterprise.elastic.co/s/skp5pl7d2f7vs

**Comment by ywelsch** [comment]
I had a look at this one: I think this exception is expected to happen in the line
```
response = getAsyncSearch(response.getId(), TimeValue.timeValueMillis(1));
```
if there's a refresh happening between the task expiry time being updated (i.e. `store.updateExpirationTime`) with now + 1 millis and the task then being retrieved from the index. The refresh makes the task visible to the `delete-by-query` of `AsyncTaskMaintenanceService`, which cleans it up, resulting in the task not being available anymore, which then results in an ResourceNotFoundException`. Note that the maintenance task is running every 100ms (see `AsyncSearchIntegTestCase.nodeSettings`).

I think that `ResourceNotFoundException` is a valid response in the test, and should be handled as such.

@jtibshirani WDYT?

**Comment by jtibshirani** [comment]
> now + 1 millis and the task then being retrieved from the index.

This is a good observation, and would explain the `ResourceNotFoundException`! It indeed looks like we first update the expiration time, then get the document, so the cleanup task could run between. This is a bit surprising to me, I figured we would use the 'get result' from the update response instead of re-fetching the document.

I don't think this explains all the errors we're seeing. In https://gradle-enterprise.elastic.co/s/6odmjpmerlk7a, the async search task is never cleaned up:
```
org.elasticsearch.xpack.search.AsyncSearchActionIT > testUpdateStoreKeepAlive FAILED
    java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([B760AFB0B7495F74:8EDD6BCC5A1AE01F]:0)
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertFalse(Assert.java:64)
        at org.junit.Assert.assertFalse(Assert.java:74)
        at org.elasticsearch.xpack.search.AsyncSearchIntegTestCase.lambda$ensureTaskRemoval$0(AsyncSearchIntegTestCase.java:198)
```

And in https://gradle-enterprise.elastic.co/s/hlbba7jpfiwm6, it looks like the +10 days update isn't reflected in the response:

```
    java.lang.AssertionError: 
    Expected: a value greater than <1638199886894L>
         but: <1638199886894L> was equal to <1638199886894L>
        at __randomizedtesting.SeedInfo.seed([BBF09564E57352EC:824D51180820ED87]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.xpack.search.AsyncSearchActionIT.testUpdateStoreKeepAlive(AsyncSearchActionIT.java:392)
```

Maybe there are a couple different issues at play?

**Comment by cbuescher** [comment]
Another one on master today: https://gradle-enterprise.elastic.co/s/jxx6bovpmm2vm

**Comment by jtibshirani** [comment]
I merged #83389 to address the `ResourceNotFoundException` errors. I don't think it will fix the other errors we've seen. I opened a new issue to track those, so we can keep this one focused on the original `ResourceNotFoundException` problem: #83422.
